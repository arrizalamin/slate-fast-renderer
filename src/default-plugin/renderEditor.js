// @flow
import * as React from 'react';
import type {RenderEditorProps} from '../plugin';
import Node from './Node';
import Editor from '../editor';

const renderEditor = (props: RenderEditorProps, editor: Editor): React.Node => {
  const {value, plugins} = props;
  const {document} = value;

  if (document && document.nodes) {
    return (
      <React.Fragment>
        {document.nodes.map((node, i) => (
          <Node
            key={i.toString()}
            editor={editor}
            node={node}
            parent={document}
            block={node}
            plugins={plugins}
          />
        ))}
      </React.Fragment>
    );
  }
  return null;
};

export default renderEditor;
