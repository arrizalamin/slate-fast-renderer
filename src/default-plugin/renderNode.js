// @flow
import * as React from 'react';
import type {RenderNodeProps} from '../plugin';

const renderNode = (props: RenderNodeProps): React.Node => props.children;

export default renderNode;
