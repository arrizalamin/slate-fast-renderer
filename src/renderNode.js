// @flow
import React from 'react';
import renderLeaf from './renderLeaf';
import type {Plugin} from './plugin';
import type {Node, Data} from './value';

const renderNode = (
  node: Node,
  index: number,
  plugins: Array<Plugin>
): ?React$Node => {
  if (node.object === 'text') {
    return (
      <React.Fragment>
        {node.leaves.map((node, index) => renderLeaf(node, index, plugins))}
      </React.Fragment>
    );
  }

  const childrenNodes =
    node.nodes &&
    node.nodes.map((node, index) => renderNode(node, index, plugins));

  const children = <React.Fragment>{childrenNodes}</React.Fragment>;

  const data: ?Data =
    node.data &&
    (node.data instanceof Map ? node.data : new Map(Object.entries(node.data)));
  const newNode = data ? Object.assign(node, {data}) : node;

  const props = {
    node: newNode,
    attributes: {index},
    children,
  };

  let i = 0;
  const next = () => {
    const plugin = plugins[i];
    i = i + 1;
    if (!plugin) {
      return null;
    }
    if (!plugin.renderNode) {
      return next();
    }

    return plugin.renderNode(props, next);
  };

  return next();
};

export default renderNode;
