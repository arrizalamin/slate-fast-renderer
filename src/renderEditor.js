// @flow
import * as React from 'react';
import type {Plugin, RenderEditorProps, NextFn} from './plugin';
import type {Value, Node, Document, Block} from './value';
import renderNode from './renderNode';
import Editor from './editor';

const renderEditor = (
  props: RenderEditorProps,
  editor: Editor,
  next: NextFn
): React.Node => {
  const {value, plugins} = props;
  const {document} = value;

  if (document && document.nodes) {
    return (
      <React.Fragment>
        {document.nodes.map(node =>
          renderNode(editor, node, document, node, plugins)
        )}
      </React.Fragment>
    );
  }
  return null;
};

export default renderEditor;
