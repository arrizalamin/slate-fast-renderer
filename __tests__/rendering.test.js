import React from 'react';
import renderer from 'react-test-renderer';
import SlateRenderer, {Value, RenderUtils} from '../src';
import renderNode from './renderNode';
import renderEditor from './renderEditor';

const valueJSON = {
  document: {
    nodes: [
      {
        object: 'block',
        type: 'paragraph',
        nodes: [
          {
            object: 'text',
            leaves: [
              {
                text: 'Hello World',
              },
            ],
          },
        ],
      },
    ],
  },
};

it('renders correctly', () => {
  const value = Value.fromJSON(valueJSON);
  const component = <SlateRenderer value={value} renderNode={renderNode} />;
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders renderEditor', () => {
  const value = Value.fromJSON(valueJSON);
  const component = (
    <SlateRenderer
      value={value}
      renderNode={renderNode}
      renderEditor={renderEditor}
    />
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders renderToArray', () => {
  const value = Value.fromJSON(valueJSON);
  const blocks = RenderUtils.renderToArray({value, renderEditor, renderNode});
  const trees = blocks.map(block => renderer.create(block).toJSON());

  expect(trees).toMatchSnapshot();
});
