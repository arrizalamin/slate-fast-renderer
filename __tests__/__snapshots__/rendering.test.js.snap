// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<p
  data-key="2"
>
  Hello World
</p>
`;

exports[`renders renderEditor 1`] = `
<div>
  <p
    data-key="7"
  >
    Hello World
  </p>
</div>
`;

exports[`renders renderToArray 1`] = `
Array [
  <p
    data-key="12"
  >
    Hello World
  </p>,
]
`;
